File: js/src/interfaces/INamespaceReader.ts
import { CustomDataLink } from './CustomDataLink';

/**
 * Interface for reading from a namespace.
 */
export interface INamespaceReader {
  /**
   * Gets the latest link with the given logical name.
   * 
   * @param logicalName The logical name to look for
   * @returns The latest link, or null if not found
   */
  getLatestAsync(logicalName: string): Promise<CustomDataLink | null>;
  
  /**
   * Streams all links newer than the given timestamp.
   * 
   * @param newerThanUnixTs Optional timestamp to filter by
   * @returns Async iterable of links
   */
  streamAsync(newerThanUnixTs?: number): AsyncIterable<CustomDataLink>;
}

File: Circles.Profiles.Interfaces/INamespaceReader.cs
using Circles.Profiles.Models;
using Circles.Profiles.Models.Core;

namespace Circles.Profiles.Interfaces;

public interface INamespaceReader
{
    /// Latest link bearing that *logical* name, or `null`.
    Task<CustomDataLink?> GetLatestAsync(string logicalName,
        CancellationToken ct = default);

    /// Streams **newest → oldest**, *already* filtered on the SDK side so that
    /// only links with `signedAt > newerThanUnixTs` are yielded.
    IAsyncEnumerable<CustomDataLink> StreamAsync(
        long newerThanUnixTs = 0,
        CancellationToken ct = default);
}